presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatStreamReader.java

```diff
@@ -0,0 +1,429 @@
+import com.facebook.presto.spi.block.VariableWidthBlockBuilder;
+import com.facebook.presto.spi.type.BigintType;
+import com.facebook.presto.spi.type.IntegerType;
+import com.facebook.presto.spi.type.MapType;
+import com.facebook.presto.spi.type.SmallintType;
+import com.facebook.presto.spi.type.TinyintType;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.io.Closer;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.joda.time.DateTimeZone;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.IN_MAP;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.PRESENT;
+import static com.facebook.presto.orc.stream.MissingInputStreamSource.missingStreamSource;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Flat Maps are a layout of maps supported in DWRF.
+ * <p>
+ * Values associated with different keys are stored in separate streams rather than having a single set of value streams for the map.
+ * <p>
+ * There is a ColumnEncoding associated with the value streams for a given key.  All the ColumnEncodings for values have the same
+ * columnId, and use a sequenceId to distinguish them.  The ColumnEncoding also contains the key it is associated with as metadata.
+ * <p>
+ * Note that the ColumnEncoding with sequenceId 0 for the values has no data associated with it, only statistics.  Similarly there
+ * is a ColumnEncoding for the key "column" which has no data associated with it, only statistics.
+ */
+public class MapFlatStreamReader
+        implements StreamReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapFlatStreamReader.class).instanceSize();
+
+    private final StreamDescriptor streamDescriptor;
+    private final DateTimeZone hiveStorageTimeZone;
+    private final AggregatedMemoryContext systemMemoryContext;
+    // This is the StreamDescriptor for the value "column" with sequence ID 0, it is used to derive StreamDescriptors for the
+    // value "column" with other sequence IDs
+    private final StreamDescriptor baseValueStreamDescriptor;
+    private final OrcType.OrcTypeKind keyOrcType;
+
+    private final List<InputStreamSource<BooleanInputStream>> inMapStreamSources = new ArrayList<>();
+    private final List<BooleanInputStream> inMapStreams = new ArrayList<>();
+    private final List<StreamReader> valueStreamReaders = new ArrayList<>();
+    private final List<StreamDescriptor> valueStreamDescriptors = new ArrayList<>();
+
+    private Block dictionaryBlock;
```

Comment:
nit: Maybe call this `keyBlockTemplate`? Also rename `constructKeyBlock` to be `getKeyBlockTemplate`, `createIntegerKeyBlock` to be `getIntegerKeyBlockTemplate`, `createSliceKeysBlock` to be `getSliceKeyBlockTemplate`